XMonad/Hooks/DynamicLog.hs

```diff
@@ -351,7 +358,36 @@ statusBarPropTo :: LayoutClass l Window
 statusBarPropTo prop cmd pp =
     makeStatusBar
         (xmonadPropLog' prop =<< dynamicLogString pp)
-        (spawnOnce cmd)
+        (spawnStatusBarAndRemember cmd)
+
+-- | Spawns a status bar and persists its PID. This is useful when the status bars
+-- should be restarted with xmonad. Use this in combination with 'cleanupStatusBars'.
+spawnStatusBarAndRemember :: String -- ^ The command used to spawn the status bar
+                          -> X()
+spawnStatusBarAndRemember cmd = do
+  newPid <- spawnPID cmd
+  XS.modify (StatusBarPIDs . (newPid :) . getPIDs)
+
+newtype StatusBarPIDs = StatusBarPIDs { getPIDs :: [ProcessID] }
+  deriving (Show, Read)
+
+instance ExtensionClass StatusBarPIDs where
+  initialValue = StatusBarPIDs []
+  extensionType = PersistentExtension
+
+
+-- | Kills the status bars started with 'spawnStatusBarAndRemember', and resets
+-- the state. This could go for example at the beginning of the startupHook.
+cleanupStatusBars :: X ()
+cleanupStatusBars =
+    getPIDs <$> (XS.get :: X StatusBarPIDs)
+    >>= (io . mapM_ killPid)
+    >> XS.put (StatusBarPIDs [])
+   where
+    killPid :: ProcessID -> IO ()
+    -- Negate the PID to kill the whole group
```

Comment:
This can be removed now that we use `signalProcessGroup`; though perhaps it would be good to add a little bit of documentation to `cleanupStatusBars` on /why/ we kill the process group, which edge cases will cause problems etc.

Reply:
@slotThe  is this better?